clippy_lints/src/fn_null_check.rs

```diff
@@ -0,0 +1,129 @@
+use clippy_utils::consts::{constant, Constant};
+use clippy_utils::diagnostics::span_lint_and_help;
+use clippy_utils::{is_integer_literal, is_path_diagnostic_item};
+use if_chain::if_chain;
+use rustc_hir::{BinOpKind, Expr, ExprKind, TyKind};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::sym;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for comparing a function pointer to null.
+    ///
+    /// ### Why is this bad?
+    /// Function pointers are assumed to not be null.
+    ///
+    /// ### Example
+    /// ```rust,ignore
+    /// let fn_ptr: fn() = /* somehow obtained nullable function pointer */
+    ///
+    /// if (fn_ptr as *const ()).is_null() { ... }
+    /// ```
+    /// Use instead:
+    /// ```rust,ignore
+    /// let fn_ptr: Option<fn()> = /* somehow obtained nullable function pointer */
+    ///
+    /// if fn_ptr.is_none() { ... }
+    /// ```
+    #[clippy::version = "1.67.0"]
+    pub FN_NULL_CHECK,
+    correctness,
```

Comment:
I would put this into `suspicious` instead of `correctness`. The code may not do what it should, but it's not dangerous in and of itself. `correctness` lints stop the build by default, so they might mask other noteworthy lint messages.

Reply:
I would argue it should stay in `correctness` honestly.

[The clippy rfc](https://github.com/rust-lang/rfcs/blob/master/text/2476-clippy-uno.md?plain=1#L87) clearly defines `correctness` as `Probable bugs`, with the example of:
> calling `.clone()` on `&&T`, which clones the (`Copy`) reference and not the actual type

Which is in my opinion of a similar logical bug category as this lint. Furthermore, [clippy itself](https://github.com/rust-lang/rust-clippy/blob/master/README.md?plain=1#L16) defines `correctness` as:
> code that is outright wrong or useless

Which is also what I think this lint is supposed to detect, checking a function pointer for null is not _just_
> code may not do what it should

It is just an outright wrong code sample. Nobody should ever want to do this, as no function pointer in rust is assumed to be nullable.